Wide-ranging updates to trust-manager docs for v0.4.0 (#1160)
cert-manager/website: content/docs/projects/trust-manager/README.md

of trust-manager from upgrading.

If you're using an official cert-manager-provided Debian CA package (which is the default), you should check which version you have
and compare against the [latest package version](https://quay.io/repository/jetstack/cert-manager-package-debian?tab=tags&tag=latest).

The version can be configured with the `.defaultPackageImage.tag` value on the Helm chart, and the version
is also written to the `status` field on any synced Bundle resource which uses the default CA package.

## Upgrading A Default CA Package Using Helm

Say we want to do an in-place upgrade of our default CA package to tagged version `XYZ` - without upgrading trust-manager.

We'll assume the Helm release is called "trust-manager" and that we've installed into the `cert-manager` namespace.

> ⚠️ This upgrade process assumes that it's the only thing running. If another user or process changes Helm values
> while you're doing this process, you might overwrite their work.

First, we'll dump our current Helm values, so we don't lose them:

```bash
helm get values -n cert-manager trust-manager -oyaml > values.yaml

> ```

Reviewer: The content of `values.yaml` will be `null` if the chart was originally installed without any custom values.
Maybe add a note about that.

It might be simpler to pass the `--reuse-values` flag instead and supply `--set defaultPackageImage.tag=XYZ` to override the last such value if it was supplied.

Author: That does seem simpler but I think I'll struggle to test that today to ensure that I'm writing sense :thinking:

Would we be OK to keep the current version of this - with the warts - and then come back to it?

Next, if `defaultPackageImage.tag` is already set in `values.yaml`, update it. Otherwise, add it.

> You can find the available tags [on `quay.io`](https://quay.io/repository/jetstack/cert-manager-package-debian?tab=tags&tag=latest).

Reviewer: I looked at that page and was confused about why the tag is a date in 2021 but the  image appears to be last modified only 18 days ago.

![image](https://user-images.githubusercontent.com/978965/216599209-ac7db19d-74f2-4ee2-b632-c4f8713d82e6.png)

I suggest explaining the significance of the date in these tags.

Author: Good idea, yeah. I'm struggling for time today but I'll try to get this done!

```yaml
# values.yaml
...
defaultPackageImage:
  tag: XYZ
```

Finally, apply back the changes, being sure to manually specify the version of trust-manager which is installed, to avoid
also updating the trust-manager controller at the same as the default CA package:

```bash
# Get the currently installed version. You could do this manually if you find that easier.
TRUST_MANAGER_VER=$(helm list --filter "^trust-manager$" -n cert-manager -ojson | jq -r ".[0].app_version")

# Check the version makes sense
echo $TRUST_MANAGER_VER

# Run the upgrade

> helm upgrade -f values.yaml -n cert-manager trust-manager jetstack/trust-manager --version $TRUST_MANAGER_VER

Reviewer: Maybe add a note to say that if you specify an incorrect `defaultPackageImage.tag`, the new deployment will fail, and you will need to use `helm rollback`.

```

## Preparing for Production

> TLS is a complicated beast and there are plenty of ways to use TLS certificates in ways which will cause problems
> down the line. We'd like to document a few of these potential gotchas here to help prepare users for running
> trust-manager safely in production.

Reviewer: I think "complicated beast" might confuse non-uk readers. Maybe instead.
```suggestion
TLS is complicated and there are many ways to misuse TLS certificates. 
Here are some potential gotchas to be aware of before running
trust-manager in production.
```

Unfortunately the existence of these files has the unfortunate side effect of people sometimes assuming that `cert.pem`
is the correct choice even when `fullchain.pem` would be correct. This means that the rest of the chain will not
be sent when the certificate is used.

Often, a quick fix that _seems_ to work for this is that clients add the chain to their trust store, which will seem
to fix certificate errors in the short term. It's easy for this kind of "fix" to end up being embedded somewhere as a
solution which others can follow.

This "fix" is dangerous; it means that the intermediate cannot be safely rotated without all trust stores
which contain it being updated first.

Intermediates in this case become _de facto_ root certificates, which completely defeats the point of having
intermediate certificates in the first place.

Avoid using intermediates in any trust store wherever possible unless you're absolutely certain they should be included.
An example of where it might be OK would be cross signing, which is not likely to be required in the general case.

It would be better to copy just the root certificate to a new `ConfigMap` and use that as a source rather than trusting

> an intermediate.

Reviewer: I find this section confusing because 

> For most users and applications `fullchain.pem` is the only correct choice.

contradicts

> Avoid using intermediates in any trust store wherever possible unless you're absolutely certain they should be included.

If the section is meant to be about what file users should instruct trust-manager to copy to the trust store.

Given the CA certificates:
 *  `Example Corp Root > Cloud Infrastructure > Linkerd`
 *  `Example Corp Root > VPN > Sales`

I would not want my linkerd client to trust a peer that presented a certificate signed by `Example Corp Root > VPN > Sales`.
So I would configure trust-manager to copy the intermediate CA  `Example Corp Root > Cloud Infrastructure > Linkerd` into the ConfigMap `linkerd/linkerd-identity-trust-roots`.

Author: > I find this section confusing because ...

That makes sense, yeah. I can see how that would be confusing.

>  So I would configure trust-manager to copy the intermediate CA Example Corp Root > Cloud Infrastructure > Linkerd into the ConfigMap linkerd/linkerd-identity-trust-roots.

This kinda thing feels like an antipattern to me. It's _so_ easy to get this wrong and it turns every single intermediate into a de facto root certificate. It would be much better to have two roots, "Corp Root Infra" and "Corp Root VPN", in this scenario.

I think orgs would be well advised not to create PKI structures like this if at all possible. Certainly I wouldn't actively recommend it, although I acknowledge that this pattern exists today.
